src/tensor/tensor.h

```diff
@@ -39,78 +61,350 @@ struct Storage {
   properties do not all hold).
 
   These properties are stricter than some other frameworks, such as PyTorch,
-  which allow the users to manually add dimensions with stride 0 (and dim>1) so
+  which allow the users to manually add dimensions with stride 0 and dim > 1 so
   that a lower-dimensional quantity can masquerade as one with a higher
-  dimension.  We require that it never be possible to access the same
-  memory location using two different tuples of indexes.  We also
-  don't allow zero dims (i.e. a tensor must not be empty); if you want an
-  empty Tensor, just use a null pointer.
+  dimension.  We require that it never be possible to access the same memory
+  location using two different tuples of indexes.  We also don't allow zero dims
+  (i.e. a tensor must not be empty); if you want an empty Tensor, just use a
+  null pointer.  In addition, require that the stride equal zero for any
+  axis that has dim = 1.
+
+  Our requirements on a TensorDim are:
 
     0 <= num_axes <= 5
     for 0 <= axis < num_axes:
        dims[i] > 0
-
-  The strides may take any value, including zero or negative, as long as the
-  uniqueness property is satisfied (i.e. must not be possible to access the
-  same memory location using two different tuples of indices.
-
+       if dims[i] = 1, then strides[i] = 0.
+       if dims[i] != 1, then strides[i] != 0
+    ... plus the uniqueness property.
+
+  The uniqueness property means that we must not be able to access
+  the same memory location via two different tuples of indexes).
+  Recause testing this property exactly would be difficult in general
+  without bringing in number theory, we test a slightly stronger version
+  of it that covers all cases we are likely to encounter.
 */
 
 struct TensorDim {
-
-  int64_t num_axes;
+  uint64_t num_axes;
   int64_t dims[KALDI_TENSOR_MAX_DIM];
   int64_t strides[KALDI_TENSOR_MAX_DIM];
   // We may later add methods to this.
 
   // Checks that the TensorDim is valid, assuming it is part of a Tensor.
-  // I.e. that it satifies the properties mentioned above.
+  // I.e. that it satifies all the properties mentioned above.
   bool Check();
 };
 
 struct TensorDimProperties {
-  // Below are cached properties that depend on a TensorDim.
+  // Below are cached properties that are derived from the underlying data in
+  // struct TensorDim.
 
   // The number of elements in the Tensor, which equals the product
-  // of dims[0] .. dims[num_axes - 1].  Must always be >0.
+  // of dims[0] .. dims[num_axes - 1].  Must be >0.
   int64_t num_elements;
 
   // is_contiguous means that the data form a contiguous block in memory; it is
-  // not the same as PyTorch's is_contiguous which is a stronger condition; our
-  // has_expected_strides is equivalent to that.
+  // not the same as PyTorch's is_contiguous which is a stronger condition,
+  // implying also that the strides are as for a `C-style` array.
   bool is_contiguous;
```

Comment:
Since this is different from pytorch is_contiguous, I think we should give it a different name to protect unsuspecting masses :)